/****************************************************************************
 *
 *   Copyright (c) 2020-2021 PX4 Development Team. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 * 3. Neither the name PX4 nor the names of its contributors may be
 *    used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
 * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 ****************************************************************************/

/**
 * @file ActuatorEffectivenessMultirotor.hpp
 *
 * Actuator effectiveness computed from rotors position and orientation
 *
 * @author Julien Lecoeur <julien.lecoeur@gmail.com>
 */

#include "ActuatorEffectivenessMultirotor.hpp"

ActuatorEffectivenessMultirotor::ActuatorEffectivenessMultirotor(ModuleParams *parent):
	ModuleParams(parent)
{
}

bool
ActuatorEffectivenessMultirotor::getEffectivenessMatrix(matrix::Matrix<float, NUM_AXES, NUM_ACTUATORS> &matrix,
		bool force)
{
	if (_updated || force) {
		_updated = false;

		// Get multirotor geometry
		MultirotorGeometry geometry{};

		for (int n = 0; n < NUM_ROTORS_MAX; n++) {
			char buffer[17];

			for (int i = 0; i < 3; i++) {
				char axis_char = 'X' + i;

				// CA_MC_Rn_P{X,Y,Z}
				sprintf(buffer, "CA_MC_R%u_P%c", n, axis_char);
				param_get(param_find(buffer), &geometry.rotors[n].position(i));

				// CA_MC_Rn_A{X,Y,Z}
				sprintf(buffer, "CA_MC_R%u_A%c", n, axis_char);
				param_get(param_find(buffer), &geometry.rotors[n].axis(i));
			}

			geometry.rotors[n].axis.normalize();

			// CA_MC_Rn_CT
			sprintf(buffer, "CA_MC_R%u_CT", n);
			param_get(param_find(buffer), &geometry.rotors[n].thrust_coef);

			// CA_MC_Rn_KM
			sprintf(buffer, "CA_MC_R%u_KM", n);
			param_get(param_find(buffer), &geometry.rotors[n].moment_ratio);
		}

		_num_actuators = computeEffectivenessMatrix(geometry, matrix);
		return true;
	}

	return false;
}

int
ActuatorEffectivenessMultirotor::computeEffectivenessMatrix(const MultirotorGeometry &geometry,
		matrix::Matrix<float, NUM_AXES, NUM_ACTUATORS> &effectiveness)
{
	int num_actuators = 0;

	effectiveness.setZero();

	for (size_t i = 0; i < NUM_ROTORS_MAX; i++) {

		// Get rotor axis
		matrix::Vector3f axis{geometry.rotors[i].axis};

		// Normalize axis
		float axis_norm = axis.norm();

		if (axis_norm > FLT_EPSILON) {
			axis /= axis_norm;

		} else {
			// Bad axis definition, ignore this rotor
			continue;
		}

		// Get rotor position
		const matrix::Vector3f position{geometry.rotors[i].position};

		// Get coefficients
		float ct = geometry.rotors[i].thrust_coef;
		float km = geometry.rotors[i].moment_ratio;

		if (fabsf(ct) < FLT_EPSILON) {
			continue;
		}

		// Compute thrust generated by this rotor
		const matrix::Vector3f thrust{ct * axis};

		// Compute moment generated by this rotor
		const matrix::Vector3f moment{ct * position.cross(axis) - ct *km * axis};

		// Fill corresponding items in effectiveness matrix
		for (size_t j = 0; j < 3; j++) {
			effectiveness(j, i) = moment(j);
			effectiveness(j + 3, i) = thrust(j);
		}

		num_actuators = i + 1;
	}

	return num_actuators;
}
